packages/aws-cdk/lib/api/deploy-stack.ts

```diff
@@ -66,34 +69,40 @@ const LARGE_TEMPLATE_SIZE_KB = 50;
 
 /** @experimental */
 export async function deployStack(options: DeployStackOptions): Promise<DeployStackResult> {
-  if (!options.stack.environment) {
-    throw new Error(`The stack ${options.stack.displayName} does not have an environment`);
+  const stack = options.stack;
+
+  if (!stack.environment) {
+    throw new Error(`The stack ${stack.displayName} does not have an environment`);
   }
 
-  const cfn = await options.sdk.cloudFormation(options.stack.environment.account, options.stack.environment.region, Mode.ForWriting);
-  const deployName = options.deployName || options.stack.stackName;
+  // Translate symbolic/unknown environment references to concrete environment references
+  const stackEnv = await options.sdk.resolveEnvironment(stack.environment.account, stack.environment.region);
+
+  const cfn = (await options.sdk.forEnvironment(stackEnv.account, stackEnv.region, Mode.ForWriting)).cloudFormation();
+  const deployName = options.deployName || stack.stackName;
 
   if (!options.force) {
+    // bail out if the current template is exactly the same as the one we are about to deploy
+    // in cdk-land, this means nothing changed because assets (and therefore nested stacks) are immutable.
     debug(`checking if we can skip this stack based on the currently deployed template and tags (use --force to override)`);
     const deployed = await getDeployedStack(cfn, deployName);
     const tagsIdentical = compareTags(deployed?.tags ?? [], options.tags ?? []);
-    if (deployed && JSON.stringify(options.stack.template) === JSON.stringify(deployed.template) && tagsIdentical) {
+    if (deployed && JSON.stringify(stack.template) === JSON.stringify(deployed.template) && tagsIdentical) {
       debug(`${deployName}: no change in template and tags, skipping (use --force to override)`);
       return {
         noOp: true,
         outputs: await getStackOutputs(cfn, deployName),
         stackArn: deployed.stackId,
-        stackArtifact: options.stack
+        stackArtifact: stack
       };
     } else {
       debug(`${deployName}: template changed, deploying...`);
     }
   }
 
-  // bail out if the current template is exactly the same as the one we are about to deploy
-  // in cdk-land, this means nothing changed because assets (and therefore nested stacks) are immutable.
+  const assets = new AssetManifestBuilder();
 
-  const params = await prepareAssets(options.stack, options.toolkitInfo, options.reuseAssets);
+  const params = await addMetadataAssetsToManifest(stack, assets, options.toolkitInfo, options.reuseAssets);
```

Comment:
I am assuming this is for backwards compatibility with stacks synthesized without an asset manifest, correct? What happens if the stack does have one?

Reply:
This is not true yet, I will add an additional `if` in the PR that introduces asset manifests from the framework side.